src/Orleans.Transactions/State/TransactionalState.cs

```diff
@@ -1,236 +1,296 @@
+                    {
+                        throw new OrleansTransactionLockAcquireTimeoutException(info.TransactionId.ToString());
+                    }
+
+                    // merge the current clock into the transaction time stamp
+                    record.Timestamp = this.queue.Clock.MergeUtcNow(info.TimeStamp);
+
+                    // link to the latest state
+                    if (record.State == null)
+                    {
+                        this.queue.GetMostRecentState(out record.State, out record.SequenceNumber);
+                    }
+
+                    // if this is the first write, make a deep copy of the state
+                    if (!record.HasCopiedState)
+                    {
+                        record.State = this.copier.DeepCopy(record.State);
+                        record.SequenceNumber++;
+                        record.HasCopiedState = true;
+                    }
+
                     if (logger.IsEnabled(LogLevel.Debug))
-                        logger.Debug($"recover two-phase-commit {pr.TransactionId}");
-                    var tm = (pr.TransactionManager == null) ? null :
-                        (ITransactionParticipant) JsonConvert.DeserializeObject<ITransactionParticipant>(pr.TransactionManager, this.serializerSettings);
+                        logger.Debug($"update-lock write v{record.SequenceNumber} {record.TransactionId} {record.Timestamp:o}");
+
+                    // record this write in the transaction info data structure
+                    info.RecordWrite(this.thisParticipant, record.Timestamp);
 
-                    commitQueue.Add(new TransactionRecord<TState>()
+                    // record this participant as a TM candidate
+                    if (info.TMCandidate != resource)
                     {
-                        Role = CommitRole.RemoteCommit,
-                        TransactionId = Guid.Parse(pr.TransactionId),
-                        Timestamp = pr.TimeStamp,
-                        State = pr.State,
-                        TransactionManager = tm,
-                        PrepareIsPersisted = true,
-                        LastSent = default(DateTime),
-                        ConfirmationResponsePromise = null
-                    });
+                        int batchsize = this.queue.BatchableOperationsCount();
+                        if (info.TMCandidate == null || batchsize > info.TMBatchSize)
+                        {
+                            info.TMCandidate = this.thisParticipant;
+                            info.TMBatchSize = batchsize;
+                        }
+                    }
+
+                    // perform the write
+                    TResult result = default(TResult);
+                    try
+                    {
+                        detectReentrancy = true;
+
+                        if (updateAction != null)
```

Comment:
This is always true, since it's guarded with an exception at the top of the method